apps/credential-service/lib/server.ts

```diff
@@ -0,0 +1,149 @@
+
+  return {
+    token,
+    revokeToken: async () => await github.apps.revokeInstallationAccessToken(),
+  };
+}
+
+/**
+ * WebSocket handler to generate temporary Github access token.
+ *
+ * The access token is automatically revoked when the websocket is closed.
+ */
+async function wsHandler(ws: WebSocket, req: IncomingMessage) {
+  /** Whether the websocket heartbeat check is still alive. */
+  let hasHeartbeat: boolean;
+  /** The interval instance for checking the heartbeat. */
+  let heartbeatInterval = setInterval(checkHeartbeat, heartBeatIntervalLength);
+  /**
+   * The repository name and owner for the request.
+   * Note: We safely case the header type as having these fields because they are checked prior
+   *       to the WebSocket handler function being invoked.
+   */
+  const {repo, owner} = req.headers as {repo: string; owner: string};
+
+  /** The temporary Github access token and function to revoke the token.. */
+  const {token, revokeToken} = await generateAccessToken(owner, repo);
+  /** Check to make sure the heartbeat is still alive. */
+  function checkHeartbeat() {
+    if (hasHeartbeat === false) {
+      ws.close(1008, 'Cannot find socket via heartbeat check');
+      ws.terminate();
+    }
+    hasHeartbeat = false;
+    ws.ping();
+  }
+
+  /**
+   * Cleans up the state of the function when the WebSocket is completed.
+   */
+  async function complete() {
+    await revokeToken();
+    clearInterval(heartbeatInterval);
+  }
+
+  // Ensure that cleanup is done when the WebSocket closes.
+  ws.on('close', complete);
+  ws.on('error', complete);
+
+  // Handle the pong response from the websocket client, updating the heartbeat as alive.
+  ws.on('pong', () => (hasHeartbeat = true));
+
+  // Send the temporary Github token to the websocket client
+  ws.send(token);
+}
+
+/**
+ * Handle upgrade requests before the websocket is used. Enforces authentication mechanisms
+ * and ensuring the required data is present in the request.
+ */
+async function upgradeHandler(req: IncomingMessage, socket: Duplex, head: Buffer) {
+  try {
+    if (!authorizationRegex.test(req.headers.authorization!)) {
+      throw Error('Missing or invalid authorization header syntax');
+    }
+    /** The NgDev token from the user to be verified. */
+    const [_, ngDevToken] = req.headers.authorization!.match(/Bearer (.*)/)!;
+    await admin
+      .auth()
+      .verifySessionCookie(ngDevToken, /* checkRevoked */ true)
+      .then((decodedToken: admin.auth.DecodedIdToken) => {
+        console.log(`Verified login of ${decodedToken.email}`);
+      });
+  } catch (e) {
+    console.error('Unable to verified authorized user');
+    console.error(e);
+    socket.write('HTTP/1.1 401 Unauthorized\r\n\r\n');
+    socket.destroy();
+    return;
+  }
+
+  /** The repository name and owner for the request. */
+  const {repo, owner} = req.headers as {repo: string; owner: string};
```

Comment:
can we make the headers more scoped to the custom use-case. Like: NgDevTargetRepoName and so on?